### PR TITLE
Implement Rewrite for [ast::Attribute]

### PR DESCRIPTION
**Pull request by cassiersg**
r? @nrc @marcusklaas

### CODE REVIEW

File: src/rewrite.rs [hunk 1]
```diff
@@ -39,4 +40,18 @@ impl<'a> RewriteContext<'a> {
             block_indent: self.block_indent + self.config.tab_spaces,
         }
     }
+
+    pub fn snippet(&self, span: Span) -> Option<String> {
```
Comment (marcusklaas) [inline src/rewrite.rs]: Should we unwrap it in here? Providing a bad span is always a programming error, correct?

File: src/rewrite.rs [hunk 2]
```diff
@@ -39,4 +40,18 @@ impl<'a> RewriteContext<'a> {
...
+        self.codemap.span_to_snippet(span).ok()
+    }
+}
+
+impl<'v> FmtVisitor<'v> {
+    pub fn mk_context(&self) -> RewriteContext<'v> {
```
Comment (marcusklaas) [inline src/rewrite.rs]: There is already a `get_context` method on `FmtVisitor`, can we use that instead?

File: src/visitor.rs [hunk 3]
```diff
@@ -402,3 +371,42 @@ impl<'a> FmtVisitor<'a> {
         }
     }
 }
+
+impl<'a> Rewrite for [ast::Attribute] {
+    fn rewrite(&self, context: &RewriteContext, _: usize, offset: usize) -> Option<String> {
+        let mut result = String::new();
+        if self.len() == 0 {
```
Comment (marcusklaas) [inline src/visitor.rs]: turbo nit: `is_empty()`

File: src/visitor.rs [hunk 4]
```diff
@@ -402,3 +371,42 @@ impl<'a> FmtVisitor<'a> {
...
+            let a_str = context.snippet(a.span).unwrap();
+
+            if i > 0 {
+                let comment = context.snippet(codemap::mk_sp(self[i-1].span.hi, a.span.lo)).unwrap();
+                // This particular horror show is to preserve line breaks in between doc
+                // comments. An alternative would be to force such line breaks to start
+                // with the usual doc comment token.
+                let multi_line = a_str.starts_with("//") && comment.matches('\n').count() > 1;
+                let comment = comment.trim();
+                if comment.len() > 0 {
```
Comment (marcusklaas) [inline src/visitor.rs]: `is_empty`?

File: src/visitor.rs [hunk 5]
```diff
@@ -402,3 +371,42 @@ impl<'a> FmtVisitor<'a> {
...
+                    result.push_str(&indent);
+                    result.push_str(comment);
```
Comment (marcusklaas) [inline src/visitor.rs]: We could rewrite the comment now as well!

**Comment by marcusklaas** [comment]
A test would be nice, maybe one where the inner comment is rewritten?

File: src/rewrite.rs (hunk 1 above)
Comment (cassiersg) [inline src/rewrite.rs]: Yes

File: src/visitor.rs (hunk 5 above)
Comment (cassiersg) [inline src/visitor.rs]: Wouldn't that transform doc comment into standard comments ?

Comment (marcusklaas) [inline src/visitor.rs]: I think not. If memory serves, a doc comment is just sugar for an attribute of the form `#[comment("Comment body")]`.

This code deals with normal comments in between the attributes.

Comment (marcusklaas) [inline src/visitor.rs]: Uhhh... I am not so sure any more now that I think about it some more.

**Comment by cassiersg** [comment]
Updated with a test. There isn't any problem to rewrite comments.
Thanks for the review, I cherry-picked this from an obsolete branch, and my review wasn't careful enough...
Is everything clear ?

**Comment by marcusklaas** [comment]
Brilliant!

r+

**Pull request merged**